lib/puppet/resource.rb

```diff
@@ -66,11 +63,35 @@ def self.from_data_hash(data, json_deserializer = nil)
     resource
   end
 
+  def self.convert_rich_data_hash(rich_data_hash, rich_data_enabled)
+    if rich_data_enabled
+      value = rich_data_hash[VALUE_KEY]
+      if value.is_a?(Array)
+        json_deserializer ||= Puppet::Pops::Serialization::Deserializer.new(Puppet::Pops::Serialization::JSON::Reader.new([]), nil)
+        reader = json_deserializer.reader
+        reader.re_initialize(value)
+        json_deserializer.read
+      else
+        pcore_type =  Puppet::Pops::Types::TypeParser.singleton.parse(rich_data_hash[PCORE_TYPE_KEY])
+        pcore_type.create(value)
+      end
+    else
+      strict = Puppet[:strict]
+      unless strict == :off
+        msg = "Unable to deserialize non-Data value for parameter #{resource.name}.#{param} unless rich data is enabled"
```

Comment:
Can you mark this for externalization? See https://github.com/puppetlabs/puppet/blob/master/CONTRIBUTING.md#writing-translatable-code for details.